File: java/core/src/test/java/dev/varar/core/SpanTest.java
package dev.varar.core;

import static org.junit.jupiter.api.Assertions.assertEquals;

import org.junit.jupiter.api.Test;

/** Port of typescript/packages/core/tests/span.test.ts. */
class SpanTest {

    @Test
    void spanFromOffsetsComputesLineAndColumnForASingleLineSource() {
        String source = "hello world";
        Span span = Span.spanFromOffsets(source, 6, 11);
        assertEquals(new Span(6, 11, 1, 7, 1, 12), span);
    }

    @Test
    void spanFromOffsetsHandlesMultiLineSources() {
        String source = "line one\nline two\nline three";
        // 'two' starts at offset 14, ends at 17
        Span span = Span.spanFromOffsets(source, 14, 17);
        assertEquals(new Span(14, 17, 2, 6, 2, 9), span);
    }

    @Test
    void spanFromOffsetsHandlesARangeCrossingANewline() {
        String source = "ab\ncd";
        // From offset 1 ('b') to 4 ('d')
        Span span = Span.spanFromOffsets(source, 1, 4);
        assertEquals(new Span(1, 4, 1, 2, 2, 2), span);
    }

    @Test
    void spanFromOffsetsHandlesAstralCharsNatively() {
        String s = "a😀b"; // 😀 is a surrogate pair: 2 UTF-16 code units
        assertEquals(4, s.length()); // UTF-16 code units, same as JS .length
        Span sp = Span.spanFromOffsets(s, 0, 4);
        assertEquals(0, sp.startOffset());
        assertEquals(4, sp.endOffset());
        assertEquals(1, sp.startLine());
        assertEquals(1, sp.startCol());
        assertEquals(1, sp.endLine());
        assertEquals(5, sp.endCol());
    }
}

File: rust/var-core/tests/span_test.rs
//! Port of `SpanTest.java` / `span.test.ts`.

use var_core::offsets::utf16_len;
use var_core::span::Span;

#[test]
fn span_from_offsets_computes_line_and_column_for_a_single_line_source() {
    let source = "hello world";
    let span = Span::from_offsets(source, 6, 11);
    assert_eq!(
        span,
        Span {
            start_offset: 6,
            end_offset: 11,
            start_line: 1,
            start_col: 7,
            end_line: 1,
            end_col: 12
        }
    );
}

#[test]
fn span_from_offsets_handles_multi_line_sources() {
    let source = "line one\nline two\nline three";
    // 'two' starts at offset 14, ends at 17
    let span = Span::from_offsets(source, 14, 17);
    assert_eq!(
        span,
        Span {
            start_offset: 14,
            end_offset: 17,
            start_line: 2,
            start_col: 6,
            end_line: 2,
            end_col: 9
        }
    );
}

#[test]
fn span_from_offsets_handles_a_range_crossing_a_newline() {
    let source = "ab\ncd";
    // From offset 1 ('b') to 4 ('d')
    let span = Span::from_offsets(source, 1, 4);
    assert_eq!(
        span,
        Span {
            start_offset: 1,
            end_offset: 4,
            start_line: 1,
            start_col: 2,
            end_line: 2,
            end_col: 2
        }
    );
}

#[test]
fn span_from_offsets_handles_astral_chars_natively() {
    let s = "a😀b"; // 😀 is a surrogate pair: 2 UTF-16 code units
    assert_eq!(4, utf16_len(s)); // UTF-16 code units, same as JS .length
    let sp = Span::from_offsets(s, 0, 4);
    assert_eq!(0, sp.start_offset);
    assert_eq!(4, sp.end_offset);
    assert_eq!(1, sp.start_line);
    assert_eq!(1, sp.start_col);
    assert_eq!(1, sp.end_line);
    assert_eq!(5, sp.end_col);
}
